clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/agent/ServiceCachingAgent.java

```diff
@@ -0,0 +1,147 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.clouddriver.ecs.provider.agent;
+
+import com.amazonaws.auth.AWSCredentialsProvider;
+import com.amazonaws.services.ecs.AmazonECS;
+import com.amazonaws.services.ecs.model.DescribeServicesRequest;
+import com.amazonaws.services.ecs.model.ListServicesRequest;
+import com.amazonaws.services.ecs.model.ListServicesResult;
+import com.amazonaws.services.ecs.model.Service;
+import com.netflix.spectator.api.Registry;
+import com.netflix.spinnaker.cats.agent.AgentDataType;
+import com.netflix.spinnaker.cats.cache.CacheData;
+import com.netflix.spinnaker.cats.cache.DefaultCacheData;
+import com.netflix.spinnaker.cats.provider.ProviderCache;
+import com.netflix.spinnaker.clouddriver.aws.security.AmazonClientProvider;
+import com.netflix.spinnaker.clouddriver.ecs.cache.Keys;
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import static com.netflix.spinnaker.cats.agent.AgentDataType.Authority.AUTHORITATIVE;
+import static com.netflix.spinnaker.cats.agent.AgentDataType.Authority.INFORMATIVE;
+import static com.netflix.spinnaker.clouddriver.ecs.cache.Keys.Namespace.ECS_CLUSTERS;
+import static com.netflix.spinnaker.clouddriver.ecs.cache.Keys.Namespace.SERVICES;
+
+public class ServiceCachingAgent extends AbstractEcsOnDemandAgent<Service> {
+  private static final Collection<AgentDataType> types = Collections.unmodifiableCollection(Arrays.asList(
+    AUTHORITATIVE.forType(SERVICES.toString()),
+    INFORMATIVE.forType(ECS_CLUSTERS.toString())
+  ));
+  private final Logger log = LoggerFactory.getLogger(getClass());
+
+  public ServiceCachingAgent(String accountName, String region, AmazonClientProvider amazonClientProvider, AWSCredentialsProvider awsCredentialsProvider, Registry registry) {
+    super(accountName, region, amazonClientProvider, awsCredentialsProvider, registry);
+  }
+
+  public static Map<String, Object> convertServiceToAttributes(String accountName, String region, Service service) {
+    Map<String, Object> attributes = new HashMap<>();
+    String applicationName = service.getServiceName().contains("-") ? StringUtils.substringBefore(service.getServiceName(), "-") : service.getServiceName();
```

Comment:
I believe this should be using Moniker instead.

Reply:
Could you please clarify this a bit.

`service` is an object from the AWS library. The service name is generally: `appname-stack-detail-version` (stack/detail being optional of course). To get the application name from the service I would still need to parse it out of the service name.

So, a bit confused by what you're suggesting.

Reply:
I'm wondering if the ECS module needs to be updated to use Moniker, rather than directly using the Frigga naming convention? If this `app-stack-detail-version` is enforced by AWS, then I guess there's nothing we can do and it's coincidental. :)

Reply:
I don't think AWS enforces it :(

Could you please provide some implementation details for Moniker, or an example/lines of code. Admittedly, I think I missed it entirely.

Reply:
Here's an example: https://github.com/spinnaker/clouddriver/blob/4fab668aef7564e8940c62b8044df84927d1239d/clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/model/LoadBalancer.java#L44

@lwander - do you have better implementation examples in CD for Moniker?

Reply:
If you want to be able to allow more flexible naming (have users define their own naming policies) you should take a look at https://github.com/spinnaker/clouddriver/blob/master/clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/v2/names/KubernetesManifestNamer.java

If not, then there is no need to do anything special

Reply:
Thank you guys for the examples!  The ECS module will need to be updated to use Moniker.

For now, can we please stick to the Frigga, and have an action to switch to Moniker right after we have a working cloud provider?

Reply:
sure.